### PR TITLE
Filter Unpublished Content

### PR DESCRIPTION
**Pull request by erikkgagewhitt**
### What does this PR do, or why is it needed?

Adds an additional condition to the LIVE_FILTER which checks that Status is equal to a number specified in the config.yml; the default number is 2 which is the `approved` status according to ROCK.

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by erikkgagewhitt** [comment]
@vinnyjth I have a couple of questions:

1. How exactly can I test this?

2. This is a little brittle as is; there's no way to specify more than one status (ie. I want published and pending content, but not denied: 1,2,!3). I don't want to take too many liberties, but is this something I could change?

**Comment by vinnyjth** [comment]
1. Let's pair on it :) 

2. You are welcome to come up with an API that would allow you to select exactly the types the user wants! I'm imagining a field in the `yml` called 

```
INCLUDE_CONTENT_TYPES
  - Approved
  - NotYetApproved
```

I'll leave it to you to come up with the implementation.

**Comment by erikkgagewhitt** [comment]
@vinnyjth I'd love to

**Comment by vinnyjth** [comment]
@erikkgagewhitt Once you tackle part 2, I'll show you how to test this.

**Comment by erikkgagewhitt** [comment]
@vinnyjth Okay, we can now have multiple allowed status, but they're not quite enumerated like I'd like them to be.

**Comment by redreceipt** [comment]
can I request that the allowed types are different per environment? I'm not sure how I could conditionally list the types in the `config.yml`

**Comment by redreceipt** [comment]
so for example, in our beta app, I want to see pending and active. gives churches a good way to preview content before it's published.

**Comment by redreceipt** [comment]
I guess a church could just include it all and then go back to take some types out, just want to kick the ball as far as we can the first time

**Comment by erikkgagewhitt** [comment]
@redreceipt I've pushed up a change that should allow you to specify Content Item status based on environment or leave it blank for default.

**Comment by redreceipt** [comment]
> @redreceipt I've pushed up a change that should allow you to specify Content Item status based on environment or leave it blank for default.

how does that work? is it comma separated or `[TYPE,TYPE]`?

**Comment by erikkgagewhitt** [comment]
Sorry, @redreceipt arrays aren't supported in env format. I'll have to come up with another solution.

**Comment by vinnyjth** [comment]
@erikkgagewhitt You might be able to fix that by wrapping this in `JSON.parse()`

https://github.com/ApollosProject/apollos-prototype/blob/master/packages/apollos-config/src/attach-env-variables.js#L15

Give it a shot! We have pretty good tests over the config module.

**Comment by erikkgagewhitt** [comment]
The most recent push _should_ allow the previous adjustment to config.yml to work. The formatting would be `SOME_ENV_VARIABLE=["string", "string"]`

**Comment by vinnyjth** [comment]
@redreceipt Can you glance at the solution to make sure it works for you? You'll need to put something like `CONTENT_TYPES=["APPROVED", "NOT_YET_APPROVED"]` in your alpha server, and `CONTENT_TYPES=["APPROVED"]` in your prod server.

**Comment by vinnyjth** [comment]
Also @redreceipt - this condition is an `AND` with the `publishedAt` dates, is that okay? So if you have content that is unapproved, but has publish dates in the future, you won't be able to see it no matter what.

**Comment by vinnyjth** [comment]
@erikkgagewhitt Testing this locally, I discovered that although the `Status` is exposed as an `Int`, you have to filter it using the `String` values. Thanks Rock! I updated your code and tests to reflect that, and I think we are good to go now!

**Comment by vinnyjth** [comment]
After @redreceipt or @richarddubay gives me a heads up that this will work for them, I will approve and merge this.

**Comment by redreceipt** [comment]
This is all apart of the content preview conversation. I know this is more work but we would love to optionally see future content as well. That's going to be the most common use case for comms, writing content for the future that they want to preview on an alpha build. 😬

**Comment by vinnyjth** [comment]
@redreceipt Would it be fine to ask the content team to not set a "published at" dates for items they want to preview?

**Comment by redreceipt** [comment]
> @redreceipt Would it be fine to ask the content team to not set a "published at" dates for items they want to preview?

We could ask them to hold off but I'd love it if we made their lives as easy as possible in that regard. I'd want them to be able to set it and forget it. Like I'm imagining them doing the content all in one sitting, date and all. And then they are free to preview the content until the start date without having to remember to go back in later and set a date. Especially when a solution doesn't seem overly complex.

If you wanna merge this one, would you be opposed to me going back in a writing a new PR to make that a toggle as well?

**Comment by richarddubay** [comment]
I think it's important that in production we don't show "unpublished" or "in-the-future content". I feel like this would be the case across churches. So I like the concept. For a potential in-app preview, being able to see the items that are "Pending Approval" in our alpha version of the app would be amazing. 

I like the idea of making the date thing a toggle. Then a church could do whatever they wanted to do. As a core team this means that we can provide tools for a church to use and we're not responsible for how they use them. We don't have to make the decision of whether or not they choose to use the tools we provide to break things. Making a decision like "no matter what anything in the future won't show" seems like a individual church based decision and not a core mandated decision ... even if I feel like most churches would choose to do so. 

@redreceipt I guess the toggle would have to be something along the lines of hide future dated things in production, but show them in development?

```
HIDE_FUTURE_DATES
   - production: true
   - development: false
```

... or something?

**Comment by redreceipt** [comment]
actually I think the implementation that would be simplest that would make the most sense is only toggling pending content and future content. So having a single env variable that would accomplish both would cover 95% of use cases:

`PREVIEW_CONTENT=true` would turn on pending and future content by default. No arrays necessary.

**Comment by vinnyjth** [comment]
@erikkgagewhitt Okay I think the feedback from @redreceipt and @richarddubay is really valuable and we need to pivot :) 

Let's orient towards adding a new `config.yml` option called something like `ROCK.SHOW_INACTIVE_CONTENT`. 

When true, we should use no filters, showing all content. When false (which should be the default) we should show only content that is Published and meets the existing `date` filters. 

Sorry for the switch up, the code in this PR is fantastic :D I should have done a bit more research before I scoped out this work.

**Comment by vinnyjth** [comment]
We can keep the JSON parsing in the `config` package :) I think that's a valuable addition that we might need in the future.

**Comment by redreceipt** [comment]
love it, thanks @vinnyjth and @erikkgagewhitt

**Comment by redreceipt** [comment]
is this one ready to review and test?

**Comment by erikkgagewhitt** [comment]
@redreceipt Should be!

**Comment by redreceipt** [comment]
did you add any test content? it's cool if not, I'll throw some in there.

**Comment by erikkgagewhitt** [comment]
I did indeed add some tests.

**Comment by redreceipt** [comment]
> I did indeed add some tests.

like test pending content in Rock?

**Comment by erikkgagewhitt** [comment]
Oh, I see. No. I only added tests around the filtering system. I didn't actually mock out any content.

**Comment by redreceipt** [comment]
something seems to be broken with pulling content now. look where my pointer is, that seems to be the culprit, the request is no longer valid.

Screen Shot 2019-10-08 at 2 34 42 PM

Screen Shot 2019-10-08 at 2 34 00 PM

**Comment by erikkgagewhitt** [comment]
Hmm, yeah; I see the issue. Not sure how that state came up; I'll take a look.

**Comment by erikkgagewhitt** [comment]
I've found the problem - going to get started on trying to fix it.

**Comment by erikkgagewhitt** [comment]
Okay, I believe I've fixed the issue with invalid requests. Updated the test to show that as well.

File: packages/apollos-data-connector-rock/src/content-items/data-source.js [hunk 1]
```diff
@@ -273,7 +273,8 @@ export default class ContentItem extends RockApolloDataSource {
       .tz(ROCK.TIMEZONE)
       .format()
       .split(/[-+]\d+:\d+/)[0];
-    return `((StartDateTime lt datetime'${date}') or (StartDateTime eq null)) and ((ExpireDateTime gt datetime'${date}') or (ExpireDateTime eq null)) `;
+    const filter = `(((StartDateTime lt datetime'${date}') or (StartDateTime eq null)) and ((ExpireDateTime gt datetime'${date}') or (ExpireDateTime eq null))) and Status eq 2}`;
```
Comment (vinnyjth) [inline packages/apollos-data-connector-rock/src/content-items/data-source.js]: I think this should be 
```suggestion
    const filter = `(((StartDateTime lt datetime'${date}') or (StartDateTime eq null)) and ((ExpireDateTime gt datetime'${date}') or (ExpireDateTime eq null))) and Status eq 'Approved'}`;
```

I found out yesterday that Rock expects a string for this field :😅

File: packages/apollos-rock-apollo-data-source/src/request-builder.js [hunk 2]
```diff
@@ -84,9 +84,11 @@ export default class RockRequestBuilder {
     return this;
   };
 
-  andFilter = (filter) => this.filter(filter, { operator: 'and' });
+  andFilter = (filter) =>
```
Comment (vinnyjth) [inline packages/apollos-rock-apollo-data-source/src/request-builder.js]: You might be able to back out your change to `andFilter` and `orFilter`, and make your improvement directly to `filter` :D

File: packages/apollos-data-connector-rock/src/content-items/data-source.js (hunk 1 above)
Comment (erikkgagewhitt) [reply packages/apollos-data-connector-rock/src/content-items/data-source.js]: I thought the values we looked at were integers - well, still strings, but strings that represent integers.

Comment (vinnyjth) [reply packages/apollos-data-connector-rock/src/content-items/data-source.js]: They are. But rock expects strings when you use the `$filter`. I only know this by playing around with the API console, try going to the `ContentChannelItems` endpoint in the Rock API browser and trying the following in the `$filter` field.

`Status eq 2`

`Status eq 'Approved'`

Comment (erikkgagewhitt) [reply packages/apollos-data-connector-rock/src/content-items/data-source.js]: Oh, okay; so `Status eq '2'` would work? Or is it preferable to use the human-readable string?

Comment (vinnyjth) [reply packages/apollos-data-connector-rock/src/content-items/data-source.js]: I don't think `Status eq '2'` would work, but you can test in the Rock API console :)

File: packages/apollos-rock-apollo-data-source/src/request-builder.js [hunk 3]
```diff
@@ -77,6 +77,9 @@ export default class RockRequestBuilder {
    * Filter resources by an odata string
    */
   filter = (filter, { operator } = { operator: 'or' }) => {
+    if (!filter) {
```
Comment (vinnyjth) [inline packages/apollos-rock-apollo-data-source/src/request-builder.js]: Will this be true with an empty string?

Comment (erikkgagewhitt) [reply packages/apollos-rock-apollo-data-source/src/request-builder.js]: I believe so. I think the empty string is falsey.

File: packages/apollos-data-connector-rock/src/content-items/data-source.js (hunk 1 above)
Comment (erikkgagewhitt) [reply packages/apollos-data-connector-rock/src/content-items/data-source.js]: Looks like it works okay, but I'm just going to leave it with the word so its clearer.

File: packages/apollos-rock-apollo-data-source/src/request-builder.js (hunk 3 above)
Comment (vinnyjth) [reply packages/apollos-rock-apollo-data-source/src/request-builder.js]: You the man! I always forget.

**Pull request merged**